lib/Widgets/SettingsSidebar/SettingsSidebar.vala

```diff
@@ -0,0 +1,97 @@
+/*
+* Copyright (c) 2017 elementary LLC. (https://elementary.io)
+*
+* This program is free software: you can redistribute it and/or modify
+* it under the terms of the GNU Lesser General Public License as published by
+* the Free Software Foundation, either version 2.1 of the License, or
+* (at your option) any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU Library General Public License for more details.
+*
+* You should have received a copy of the GNU Lesser General Public
+* License along with this program; if not, write to the
+* Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+* Boston, MA 02110-1301 USA
+*/
+
+/**
+ * SettingsSidebar acts as a controller for a Gtk.Stack; it shows a row of buttons
+ * to switch between the various pages of the associated stack widget.
+ *
+ * All the content for the rows comes from the child properties of a Granite.SettingsPage
+ * inside of the Gtk.Stack
+ *
+ */
+public class Granite.SettingsSidebar : Gtk.ScrolledWindow {
+    /**
+     * The Gtk.Stack to control
+     */
+    public Gtk.Stack stack { get; construct; }
+
+    /**
+     * Create a new SettingsSidebar
+     */
+    public SettingsSidebar (Gtk.Stack stack) {
+        Object (
+            hscrollbar_policy: Gtk.PolicyType.NEVER,
+            stack: stack,
+            width_request: 200
+        );
+    }
+
+    construct {
+        var listbox = new Gtk.ListBox ();
+        listbox.activate_on_single_click = true;
+        listbox.selection_mode = Gtk.SelectionMode.SINGLE;
+
+        add (listbox);
+
+        foreach (unowned Gtk.Widget child in stack.get_children ()) {
+            string name;
+
+            stack.child_get (child, "name", out name, null);
+
+            var page = (SettingsPage) child;
```

Comment:
The `child` is not guaranteed to be `SettingsPage` hence this can throw errors or even crash. You should check if the `child` is actually a type of `SettingsPage`.